File: paper-home/scripts/cutout-white.py
"""Key out near-white backgrounds from line-art JPGs and crop."""
import os
import sys

sys.path.insert(0, r"C:\Users\19045\anaconda3\envs\teacher\Lib\site-packages")
from PIL import Image
import numpy as np

SRC = r"D:\bysq_D"
DST = r"D:\bysq\paper-home\assets\sprites"

JOBS = [
    ("Pt8DW.jpg", "coast.png"),
    ("yun1.jpg", "cloud-1.png"),
    ("yun2.jpg", "cloud-2.png"),
    ("yun3.jpg", "cloud-3.png"),
]


def cutout(src_path, dst_path):
    im = Image.open(src_path).convert("RGBA")
    arr = np.asarray(im).astype(np.float32)
    rgb = arr[..., :3] / 255.0
    luma = rgb[..., 0] * 0.299 + rgb[..., 1] * 0.587 + rgb[..., 2] * 0.114
    ink = 1.0 - luma
    alpha = np.clip((ink - 0.04) / 0.22, 0.0, 1.0)
    out = np.zeros_like(arr)
    out[..., 0] = 42
    out[..., 1] = 38
    out[..., 2] = 34
    out[..., 3] = alpha * 255.0
    img = Image.fromarray(out.astype(np.uint8), "RGBA")
    bbox = img.getbbox()
    if bbox:
        pad = 8
        x0, y0, x1, y1 = bbox
        x0 = max(0, x0 - pad)
        y0 = max(0, y0 - pad)
        x1 = min(img.width, x1 + pad)
        y1 = min(img.height, y1 + pad)
        img = img.crop((x0, y0, x1, y1))
    img.save(dst_path, "PNG")
    print("wrote", dst_path, img.size)


def main():
    os.makedirs(DST, exist_ok=True)
    for src_name, dst_name in JOBS:
        cutout(os.path.join(SRC, src_name), os.path.join(DST, dst_name))


if __name__ == "__main__":
    main()

File: paper-home/scripts/cutout-white.mjs
import fs from 'node:fs';
import path from 'node:path';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
let sharp;
try {
  sharp = (await import('sharp')).default;
} catch {
  sharp = (await import('D:/bysq/.glbtools/node_modules/sharp/lib/index.js')).default;
}

const SRC = 'D:/bysq_D';
const DST = 'D:/bysq/paper-home/assets/sprites';

const JOBS = [
  ['Pt8DW.jpg', 'coast.png'],
  ['yun1.jpg', 'cloud-1.png'],
  ['yun2.jpg', 'cloud-2.png'],
  ['yun3.jpg', 'cloud-3.png'],
];

fs.mkdirSync(DST, { recursive: true });

for (const [srcName, dstName] of JOBS) {
  const src = path.join(SRC, srcName);
  const { data, info } = await sharp(src)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });
  const w = info.width;
  const h = info.height;
  const out = Buffer.alloc(w * h * 4);
  for (let i = 0; i < w * h; i++) {
    const r = data[i * 4];
    const g = data[i * 4 + 1];
    const b = data[i * 4 + 2];
    const luma = (r * 0.299 + g * 0.587 + b * 0.114) / 255;
    const ink = 1 - luma;
    const a = Math.max(0, Math.min(1, (ink - 0.04) / 0.22));
    out[i * 4] = 42;
    out[i * 4 + 1] = 38;
    out[i * 4 + 2] = 34;
    out[i * 4 + 3] = Math.round(a * 255);
  }
  let minX = w, minY = h, maxX = 0, maxY = 0;
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (out[(y * w + x) * 4 + 3] > 8) {
        if (x < minX) minX = x;
        if (y < minY) minY = y;
        if (x > maxX) maxX = x;
        if (y > maxY) maxY = y;
      }
    }
  }
  const pad = 8;
  minX = Math.max(0, minX - pad);
  minY = Math.max(0, minY - pad);
  maxX = Math.min(w - 1, maxX + pad);
  maxY = Math.min(h - 1, maxY + pad);
  const cw = maxX - minX + 1;
  const ch = maxY - minY + 1;
  const crop = Buffer.alloc(cw * ch * 4);
  for (let y = 0; y < ch; y++) {
    out.copy(
      crop,
      y * cw * 4,
      ((minY + y) * w + minX) * 4,
      ((minY + y) * w + minX + cw) * 4
    );
  }
  const dst = path.join(DST, dstName);
  await sharp(crop, { raw: { width: cw, height: ch, channels: 4 } }).png().toFile(dst);
  console.log('wrote', dst, cw, ch);
}
